### PR TITLE
SDKS-2621: Save counters

### PR DESCRIPTION
**Pull request by javrudsky**
# Android SDK

## What did you accomplish?
Added:
- Impression counter storage
- Impression saving counter task
- Tests
Refactored and created base persistent storage component for recorded data

### CODE REVIEW

File: src/main/java/io/split/android/client/service/synchronizer/SynchronizerImpl.java [hunk 1]
```diff
@@ -180,8 +182,11 @@ private void setupListeners() {
     }
 
     public void pause() {
-        mTaskExecutor.pause();
+        // saveImpressionsCount task is sent on pause, but it could be executed
+        // when task executor resumes when app becomes active
+        //saveImpressionsCount();
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/service/synchronizer/SynchronizerImpl.java:187]: shouldn't this line be uncomment? if not, `saveImpressionsCount` is never used.

File: src/main/java/io/split/android/client/storage/db/ImpressionsCountEntity.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    @PrimaryKey(autoGenerate = true)
+    private long id;
+
+    @NonNull
+    private String body;
+
+    @ColumnInfo(name = "created_at")
+    private long createdAt;
+    private int status;
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/storage/db/ImpressionsCountEntity.java:21]: New line in 21?

File: src/main/java/io/split/android/client/storage/impressions/SqLitePersistentImpressionsCountStorage.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    protected void insert(@NonNull ImpressionsCountEntity entity) {
+        mDao.insert(entity);
+    }
+
+    @Override
+    protected void insert(@NonNull List<ImpressionsCountEntity> entities) {
+        mDao.insert(entities);
+    }
+
+    @NonNull
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/storage/impressions/SqLitePersistentImpressionsCountStorage.java:44]: duplicated @NonNull annotation

**Review by EmilianoSanchez — APPROVED** [review]
Approved Javi, but there are a few comments to check.

File: src/main/java/io/split/android/client/service/synchronizer/SynchronizerImpl.java (hunk 1 above)
Comment (javrudsky) [reply src/main/java/io/split/android/client/service/synchronizer/SynchronizerImpl.java:187]: Yup doc. This going to be uncommented in next PR

**Pull request merged**